Add the possibility to cancel a game

krystofNovotny opened the pull request:
Games can now be cancelled by user who created them.

Implementation notes:
Cancelling is realized by `deleted` flag. Games flagged as deleted are not propagated into `player_games` file and `open_games` file, therefore are not returned to be displayed.

CreatedBy property was added to `GameSummary`, so frontend can hide the delete button for users without the right to delete the game. 

Data migration is not handled - it won't be possible to cancel games which are already created (because CreatedBy attribute is missing from the summary). I don't think it matters too much and I'm also quite not sure how to do it properly.

pongstylin commented:
Very good first effort playing in my sandbox.  I checked it out in my dev environment and played with it a bit.  I also looked over the code changes.  I have a few requests:

1. Since I can not think of any value in keeping these games around since they haven't even started yet, I prefer to hard delete them rather than soft delete them.  This will require a new FileAdapter method: cancelOpenGame().  This method can do whatever it wants, but I suggest it call a "_deleteFile" method that follows the same pattern as the "_createFile" method.  It also needs to remove the game from the open and player game files.  It is easy to run into locking-related problems, so expect me to provide more guidance on that after I review your changes.
2. I expect that cancelling games will be the exception, not the rule.  So, I would like to avoid cluttering up the game list interface with a cancel button that shows up on some games (yours), but not others (theirs).  Rather, let's force the user to click the open game in question and cancel it on the game page.

A happy side-effect of these suggestions is that you won't need to worry about migrations.  You can cancel any open game that belongs to you no matter when it was created.

krystofNovotny commented:
I have updated the commit according to your suggestions.

pongstylin commented:
Alright, as promised, I'm going to provide some guidance on how to use locking to solve some difficult problems.  Here is a summary of the steps that your cancelGame() adapter method takes:
1) Read player ID (only one expected) from the game.
2) Read the player summary list.
3) Write the player summary list with the game omitted.
4) Read the open game summary list.
5) Write the open game summary list with the game omitted. (only necessary for public games)
6) Delete the game file.

The problem with these steps is that they aren't atomic, i.e. a single transaction.  The fact that JavaScript is single-threaded can help protect you.  In other words, 2 things can't happen at the same time.  But because of asynchronous promises, you can still suffer from cases where code can pause execution to allow other things to happen before the code resumes.  Here are a couple examples of how data changing between these steps can cause problems:

1) Code pauses while reading the player summary list in step 2.  Anything might have happened.  Perhaps somebody joined the game.  The call to "onJoinGameRequest" returns successfully and the game now shows up in the player's game summary list.  But every time they click the game, it says it doesn't exist.  Your code had already only found one player ID in step 1 and would not have removed it from the 2nd player's game summary.
2) Code paused between reading the open game summary list in step 4 and writing it in step 5.  Let's say between reading and writing somebody created a new open game.  Now your code will overwrite the file and lose track of the game created between reading and writing this list.  The player wonders why their game just disappeared.

Let me know if you don't understand these examples and would like a more detailed description.  To solve these problems, we need to lock a file BEFORE reading it and release the lock AFTER writing it.  This guarantees that the file won't change between reading and writing it.  If your code pauses and other code wants to write to the file, then they will first try to get a lock and that lock will wait until the existing lock is released.  Here's a description of some better steps:
1) Get a 'write' lock on the game file since you intend to delete the file.
2) After a lock is obtained, read the file.
3) Make sure the game still only has one team.  (this is a better test than checking if it had started since joining a game and starting a game are separate steps).
4) Abort everything if the game has more than one team.
5) Use FileAdapter::_lockAndUpdateFile() to modify the player game summary as a single transaction (lock this file before reading and release after writing).
6) Use FileAdapter::_lockAndUpdateFile() to modify the open game summary as a single transaction (lock this file before reading and release after writing).
7) Delete the game file.
8) Release the lock on the game file.  This is automatic.  My lock implementation uses a callback function to do stuff between obtaining and releasing a lock.

krystofNovotny commented:
> Make sure the game still only has one team. (this is a better test than checking if it had started since joining a game and starting a game are separate steps).

In case of a game for more than 2 players, shouldn't it be possible to cancel the game and kick all the players out if it hasn't started yet?

pongstylin commented:
> > Make sure the game still only has one team. (this is a better test than checking if it had started since joining a game and starting a game are separate steps).
> 
> In case of a game for more than 2 players, shouldn't it be possible to cancel the game and kick all the players out if it hasn't started yet?

Yeah, ok.  I'm persuaded.

krystofNovotny commented:
I have reworked the `FileAdapter::cancelGame` method to properly lock files it modifies / deletes.

I have also changed the `GameService::onCancelGameRequest` to remove all the players from game groups.

pongstylin commented on src/data/FileAdapter.js:
```diff
@@ -108,6 +108,35 @@ export default class {
+        gameRefreshed.state.teams.filter(t => t && !!t.playerId).map(t => t.playerId)
+      );
+
+      [...playerIds].forEach(playerId =>
```
Be sure to map and await all promises to make sure these transactions are complete before releasing the lock (returning from the function.

krystofNovotny replied:
done

pongstylin commented on src/data/FileAdapter.js:
```diff
@@ -108,6 +108,35 @@ export default class {
+        summaries.delete(gameRefreshed.id);
+        return summaries;
+      })
+      this._deleteFile(`game_${game.id}`);
```
Await this promise as well.

krystofNovotny replied:
done

pongstylin commented on src/data/FileAdapter.js:
```diff
@@ -108,6 +108,37 @@ export default class {
+
+    return this._lock(`game_${game.id}`, 'write', async () => {
+      let gameRefreshed = await this._readFile(`game_${game.id}`, null);
+      if (gameRefreshed.state.started) {
```
If, somehow, the client manages to submit 2 near-parallel cancel game requests for the same game, then I expect one of them will call this method first, and the other one will wait for a lock.  The 2nd call will then attempt to read the file.  But the file doesn't exist.  It won't throw an error because you passed an initial value of NULL.  This is good.  You don't want to throw an exception if the file doesn't exist.  But it will, because you are assuming "gameRefreshed" has a value.  Instead, I think the right thing to do would be to return "true" if gameRefreshed is null.  This makes the cancel/delete operation idempotent.

krystofNovotny replied:
Added return true if gameRefreshed is null, also added check to GameService to make sure it doesn't try to delete gamePara which has already might have already been deleted by the first request.

pongstylin commented:
Good enough!  The only issues left that I can see are mere quibbles that I can resolve if they become something more than that.  I'm comfortable releasing this to the beta server immediately for eventual release to the playground server with all the other beta work in June.  Thanks!

By the way, if you have joined the TAO discord, let me know your name so that I can properly credit you with the work.

krystofNovotny commented:
Nice, I'm happy to help. 

I have not joined the discord, but I would like to. Could you send me a link? I'm having trouble finding it.

pongstylin commented:
A link is actually on the home page of the site.  Here's the beta server link for convenience.
https://tactics-beta.taorankings.com/
